subsys/bluetooth/audio/ascs.c

```diff
@@ -960,6 +967,12 @@ static void ascs_ep_iso_disconnected(struct bt_bap_ep *ep, uint8_t reason)
 
 	LOG_DBG("stream %p ep %p reason 0x%02x", stream, stream->ep, reason);
 
+	stream_ops = stream->ops;
+	if (stream_ops != NULL && stream_ops->disconnected != NULL) {
+		LOG_ERR("EXPECT DISCONNECT %p", stream_ops->disconnected);
+		stream_ops->disconnected(stream, reason);
```

Comment:
API description mentioned: 

"If the stream shares an isochronous channel with another stream, then disconnected callback may not be called, even if the stream is leaving the streaming state".   Implementation seems not take into account if there is another stream which is sharing the ISO channel.

Reply:
Can you elaborate? This is done in the `ascs_ep_iso_disconnected` function, so regardless of the state of the stream and the state of any paired streams, the CIS is disconnected here.

Reply:
> does it mean that only the last stream attached to a shared ISO leaving stream state (stopped_cb) will call disconnected callback (bt_bap_stream_ops) ?

Consider the following:

Streams A and B share the same CIS. 

Stream A is in the enabling state, and Stream B is in the QoS configured state. 
The CIS is connected, which triggers the `connected` callback on both streams. 
Stream A enters the streaming state. 
Stream B is later configured to enter the streaming state as well (which does not affect the CIS). 

Stream A then leaves the streaming state, and goes into the QoS Configured state, but Stream B is still in the streaming state. This does not trigger any `disconnected` callbacks. 

Stream A is then released, and the memory is free'd. This does not affect the CIS. 

Finally the CIS gets disconnected, and here only Stream B will get the `disconnected` callback. 

It's a bit confusing IMO, but that is what happens when an object (the CIS) can be used by multiple other objects (the BAP streams), and where there is a somewhat loose coupling between them.

Reply:
thanks for thorough explanation :-)

Reply:
@pin-zephyr Let me know if you think the documentation needs to be improved

Reply:
thank you.  "If the stream shares an isochronous channel with another stream, then disconnected callback may not be called, even if the stream is released" , I will have less confusion.  :-)

Reply:
> even if the stream is released

Is not really accurate though. Typically, if a stream isn't share a CIS with another stream, the CIS will be disconnected when leaving the streaming state, and not just when it is released. 

So if a stream goes from streaming to QoS Configured/Codec Configured/Idle state, then the CIS is disconnected *unless* it shares a CIS with another stream still in the streaming state.

Reply:
LGTM